extract_features.py

```diff
@@ -0,0 +1,176 @@
+
+def _convert_examples_to_features(examples, label_list, max_seq_length, tokenizer):
+    """ convert each example in a list to an InputFeatures """
+    InputFeatures = namedtuple("InputFeatures", ["input_ids", "input_mask", "segment_ids", "label_id", "tokens", "case_id"])
+    
+    labels_to_int = {label:i for i,label in enumerate(label_list)}
+    print(labels_to_int)
+    
+    features = []
+    for (ex_index, example) in tqdm(enumerate(examples), desc="converting examples"):
+        tokens = tokenizer.tokenize(example.text)
+        if len(tokens) > max_seq_length - 2 :
+             tokens = tokens[:(max_seq_length - 2)]
+        tokens = ["[CLS]"] + tokens + ["[SEP]"]
+        segment_ids = [0] * len(tokens)
+        input_ids = tokenizer.convert_tokens_to_ids(tokens)
+
+        # The mask has 1 for real tokens and 0 for padding tokens. Only real
+        # tokens are attended to.
+        input_mask = [1] * len(input_ids)
+
+        # Zero-pad up to the sequence length.
+        padding = [0] * (max_seq_length - len(input_ids))
+        input_ids += padding
+        input_mask += padding
+        segment_ids += padding
+        
+
+        assert len(input_ids) == max_seq_length
+        assert len(input_mask) == max_seq_length
+        assert len(segment_ids) == max_seq_length
+
+        label_id = labels_to_int[example.label]
+        case_id = example.guid
+        features.append(InputFeatures(input_ids, input_mask, segment_ids, label_id, tokens, case_id))
+    return features
+
+def get_args():
+    parser = argparse.ArgumentParser()
+    parser.add_argument("--num_examples", type=int, default=10000,
+        help="number of examples from which features will be extracted"
+    )
+    parser.add_argument("--output_dir", type=str, default="./output",
+        help="place to write features.json"
+    )
+    parser.add_argument("--num_labels", type=int, default=2,
+        help="number of classification labels"
+    )
+    parser.add_argument("--max_seq_length", type=int, default=128,
+        help="max length of a sequence"
+    )
+    parser.add_argument("--cased", action="store_true",
+        help="whether to consider letter case"
+    )
+    parser.add_argument("--bert_model", type=str, default="bert-base-uncased",
+        help="the type of pretrained bert model (default: 'bert-base-uncased')"
+    )
+    parser.add_argument("--trained_model_dir", type=str, default="./output",
+        help="directory storing pytorch_model.bin and bert_config.json files to load model"
+    )
+    parser.add_argument("--seed", type=int, default=42,
+        help="random seed"
+    )
+    parser.add_argument("--device", type=str, default="cuda",
+        help="where to perform calculations. 'cpu' or 'cuda'"
+    )
+    parser.add_argument("--data_dir", type=str, required=True,
+        help="location of stored data. Must have a test.csv"
+    )
+    return parser.parse_args()
+    
+
+if __name__ == "__main__":
```

Comment:
Also a very long function that would benefit from inline comments and/or being extracted into smaller functions. (I might simply not bother with extracting variables from `args` and instead just pass `args` around and let functions get what they needed from it, and then have the rest of this section read something like
```
build_model()
predict_labels()
build_feature_dict()
model.eval()
label_predictor.eval()
with open(output_file, "w", encoding='utf-8') as writer:
    do_something()
    do_something_else()
```

(I'm not clear enough on what's happening in the for loop to know what those last functions should be called; also obviously some of these might need to return stuff and pass data, but you get the overall point -- writing it this way would let me read what's happening in this function in outline form rather than essay form, and then I could dip down into the functions for details as needed.)